lib/proxy/client.go

```diff
@@ -0,0 +1,505 @@
+type Client struct {
+	sync.RWMutex
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config  ClientConfig
+	conns   map[string]*clientConn
+	metrics *clientMetrics
+}
+
+type connections struct{}
+
+// NewClient creats a new peer proxy client.
+func NewClient(config ClientConfig) (*Client, error) {
+	err := config.checkAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	metrics, err := newClientMetrics()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	closeContext, cancel := context.WithCancel(config.Context)
+
+	c := &Client{
+		config:  config,
+		ctx:     closeContext,
+		cancel:  cancel,
+		conns:   make(map[string]*clientConn),
+		metrics: metrics,
+	}
+
+	if c.config.sync != nil {
+		go c.config.sync()
+	} else {
+		go c.sync()
+	}
+
+	return c, nil
+}
+
+// sync runs the proxy watcher and cleanup functionalities.
+func (c *Client) sync() {
+	ticker := c.config.Clock.NewTicker(defaults.ResyncInterval)
+	defer ticker.Stop()
+
+	proxyWatcher, err := services.NewProxyWatcher(c.ctx, services.ProxyWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.Component(teleport.ComponentProxyPeer),
+			Client:    c.config.AccessPoint,
+			Log:       c.config.Log,
+		},
+	})
+	if err != nil {
+		c.config.Log.Errorf("Error initializing proxy peer watcher: %+v.", err)
+		return
+	}
+	defer proxyWatcher.Close()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			c.config.Log.Debug("Stopping peer proxy sync.")
+			return
+		case proxies := <-proxyWatcher.ProxiesC:
+			if err := c.updateConnections(proxies); err != nil {
+				c.config.Log.Errorf("Error syncing peer proxies: %+v.", err)
+			}
+		case <-ticker.Chan():
+			if err := c.cleanup(); err != nil {
+				c.config.Log.Errorf("Error cleaning up peer proxies: %+v.", err)
+			}
+		}
+	}
+}
+
+func (c *Client) updateConnections(proxies []types.Server) error {
+	c.Lock()
+	defer c.Unlock()
+	var errs []error
+
+	toDial := make(map[string]types.Server)
+	for _, proxy := range proxies {
+		toDial[proxy.GetName()] = proxy
+	}
+
+	for id, conn := range c.conns {
+		_, ok := toDial[id]
+
+		// delete nonexistent connections
+		if !ok {
+			c.closeConn(conn)
+			continue
+		}
+
+		// test existing connections
+		if err := c.testConnection(conn); err != nil {
+			errs = append(errs, err)
+			c.closeConn(conn)
+			continue
+		}
+	}
+
+	for id, proxy := range toDial {
+		// skips itself
+		if id == c.config.ID {
+			continue
+		}
+
+		// skip existing connection. they've been tested above.
+		if _, ok := c.conns[id]; ok {
+			continue
+		}
+
+		// establish new connections
+		conn, err := c.connect(id, proxy.GetPeerAddr())
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := c.testConnection(conn); err != nil {
+			errs = append(errs, err)
+			c.closeConn(conn)
+			continue
+		}
+
+		c.conns[id] = conn
+	}
+
+	return trace.NewAggregate(errs...)
+}
+
+func (c *Client) cleanup() error {
```

Comment:
Do we need this extra cleanup step? The connection cleanup seems to be handled in `updateConnections`. Also as I mentioned above event if `testConnection` fails the `grpc.ClientConn` will handle reconnecting a transient failure.

Reply:
It takes a long time for the watcher to notice a proxy is no longer available. But it does a good job of noticing newly added proxies. At least that's what I saw during testing. 
So the only real benefit for it is metrics, as the alternative would be to fail when Dial is called.

Reply:
I think this is due to the the ttl set in the cache/backend. Once a proxy is disconnected it stays around in the cache/backend for something like 10-20 minutes.

What is the problem, that the `tunnelErrorCounter` would be lower until the cache is updated and the client gets closed? 

We can't really be sure whether its a transient failure or the proxy was permanently disconnected until some arbitrary timeout anyways so why not use the cache/backend as the source of truth for this?

And this relates to your comment above...

> as for the connection testing, the connection state isn't always up to date and we need to do something with it for GetState to return the correct one, that's why testConnection is there. Maybe I should check the error returned or the state right after and keep the connection in case it's a transient failure and not something more permanent.

If we wait for the proxy to expire in the backend/cache we don't need to worry about checking the state. And when we are in this state where the proxy is permanently disconnected but still in the cache, I think the failure would happen when establishing/sending on the stream. Maybe it would show as a timeout